src/main/java/com/streamr/client/authentication/AuthenticationMethod.java

```diff
@@ -40,25 +44,28 @@ public String newSessionToken() {
      */
     protected abstract LoginResponse login() throws IOException;
 
-    protected LoginResponse post(String endpoint, String requestBody) throws IOException {
+    protected BufferedSource post(String endpoint, String requestBody) throws IOException {
         OkHttpClient client = new OkHttpClient();
 
         Request request = new Request.Builder()
-                        .url(options.getRestApiUrl() + endpoint)
+                        .url(restApiUrl + endpoint)
                         .post(RequestBody.create(HttpUtils.jsonType, requestBody))
                         .build();
 
         // Execute the request and retrieve the response.
         Response response = client.newCall(request).execute();
 
-        // TODO: remove this hack once the API returns 401 on invalid credentials instead of 400
-        if (response.code() == 400) {
+        if (response.code() == 401) {
```

Comment:
This whole `if` should be unnecessary now, as in `HttpUtils.assertSuccessful(response)` the appropriate exception is thrown based on status code:

```
            switch (response.code()) {
                case HttpURLConnection.HTTP_NOT_FOUND:
                    throw new ResourceNotFoundException(action);
                case HttpURLConnection.HTTP_UNAUTHORIZED:
                    throw new AuthenticationException(action);
                case HttpURLConnection.HTTP_FORBIDDEN:
                    throw new PermissionDeniedException(action);
                default:
                    throw new RuntimeException(action
                            + " failed with HTTP status "
                            + response.code()
                            + ":" + response.body().string());
            }
```